api/src/test/java/org/pmiops/workbench/monitoring/LogsBasedMetricsServiceTest.java

```diff
@@ -121,12 +132,103 @@ public void testRecord_handlesMultipleMeasurements() {
         .containsAllIn(
             ImmutableSet.of(
                 EventMetric.NOTEBOOK_CLONE.getName(), EventMetric.NOTEBOOK_DELETE.getName()));
   }
+
+  @Test
+  public void testTimeAndRecordWithRunnable() {
+    Set<Integer> sideEffectSet = new HashSet<>();
+    logsBasedMetricService.recordElapsedTime(
+        MeasurementBundle.builder().addTag(MetricLabel.OPERATION_NAME, "test1"),
+        DistributionMetric.WORKSPACE_OPERATION_TIME,
+        () -> {
+          int innerInt = 2;
+          sideEffectSet.add(3);
+          innerInt -= 3;
+          assertThat(innerInt).isEqualTo(-1);
+        });
+    assertThat(sideEffectSet).contains(3);
+    verify(mockLogging).write(logEntriesCaptor.capture());
+    final Map<String, Object> entryData =
+        StreamSupport.stream(logEntriesCaptor.getValue().spliterator(), false)
+            .map(LogEntry::getPayload)
+            .map(p -> (JsonPayload) p)
+            .map(JsonPayload::getDataAsMap)
+            .findFirst()
+            .orElse(Collections.emptyMap());
+    assertThat(entryData).hasSize(4);
+    assertThat(entryData.get(PayloadKey.NAME.getKeyName()))
+        .isEqualTo(DistributionMetric.WORKSPACE_OPERATION_TIME.getName());
+    assertThat((double) entryData.get(PayloadKey.VALUE.getKeyName()))
+        .isEqualTo((double) OPERATION_DURATION.toMillis());
+  }
+
+  public void testTimeAndRecordWithSupplier() {
+    Set<Integer> aSet = new HashSet<>();
+    final int result =
+        logsBasedMetricService.recordElapsedTime(
+            MeasurementBundle.builder().addTag(MetricLabel.OPERATION_NAME, "test1"),
+            DistributionMetric.WORKSPACE_OPERATION_TIME,
+            () -> {
+              return 99;
+            });
+    assertThat(aSet).contains(3);
+    assertThat(result).isEqualTo(99);
+
+    verify(mockLogging).write(logEntriesCaptor.capture());
+    final Map<String, Object> entryData =
+        StreamSupport.stream(logEntriesCaptor.getValue().spliterator(), false)
+            .map(LogEntry::getPayload)
+            .map(p -> (JsonPayload) p)
+            .map(JsonPayload::getDataAsMap)
+            .findFirst()
+            .orElse(Collections.emptyMap());
+    assertThat(entryData).hasSize(4);
+    assertThat(entryData.get(PayloadKey.NAME.getKeyName()))
+        .isEqualTo(DistributionMetric.WORKSPACE_OPERATION_TIME.getName());
+    assertThat(entryData.get(PayloadKey.UNIT.getKeyName()))
+        .isEqualTo(UnitOfMeasure.MILLISECOND.getUcmSymbol());
+
+    assertThat((Double) entryData.get(PayloadKey.VALUE.getKeyName()))
+        .isEqualTo(OPERATION_DURATION.toMillis());
+  }
+
+  @Test(expected = IllegalAccessError.class)
+  public void testRecordElapsedTime_throws() {
+    logsBasedMetricService.recordElapsedTime(
+        MeasurementBundle.builder(),
+        DistributionMetric.COHORT_OPERATION_TIME,
+        () -> {
+          throw new IllegalAccessError("Boo!");
+        });
+  }
+
+  @Test
+  public void testRecordElapsedTime_nestedWorks() {
```

Comment:
worth making some asserts on the logs here?

Reply:
Not sure if I can mock the stopwatch to give ttwo different times.  I'll see what I can do.